### PR TITLE
rustc_span: Generate keyword classification functions automatically

### PR DESCRIPTION
**Pull request by petrochenkov**
The generated functions are compatible with symbols not forming continuous ranges.
This is a nicer alternative to https://github.com/rust-lang/rust/pull/74554/commits/48966144d4ad7d409152a405633acfc5924ec8cc from https://github.com/rust-lang/rust/pull/74554.
r? @nnethercote

### CODE REVIEW

**Comment by petrochenkov** [comment]
@bors try @rust-timer queue

The classification functions are array searches rather than simpler range checks, so they are supposed to be slower.
However, parsing is fast in general, so this shouldn't have a noticeable effect?

For small lengths (like those of the keyword lists) linear search is preferable to binary search.
For optimal speed the arrays should be sorted (to please branch predictor), but this would require copypasting some logic from `rustc_span` to `rustc_macros` after https://github.com/rust-lang/rust/pull/74554 is merged.

File: src/librustc_macros/src/symbols.rs [hunk 1]
```diff
@@ -106,18 +128,31 @@ pub fn symbols(input: TokenStream) -> TokenStream {
...
+            });
+            class_stream.extend(quote! {
+                kw::#name,
+            });
+            counter += 1;
+        }
+        if let Some(class) = class {
+            keyword_class_stream.extend(quote! {
+                fn #class(self) -> bool {
+                    [#class_stream].contains(&self)
```
Comment (Mark-Simulacrum) [inline src/librustc_macros/src/symbols.rs:152]: It does seem a bit unfortunate to lose the O(1) range-checking here, but I guess this is probably not too hot.

Maybe LLVM takes care of things for us, too, though I somewhat doubt it.

**Comment by Mark-Simulacrum** [comment]
Seems like a definite slow down, but mostly only affecting doc builds (1-2% on smaller crates). <1% regression in instruction counts on non-doc builds.

**Comment by nnethercote** [comment]
Yeah, maybe we shouldn't land this until we are certain about #74554?

File: src/librustc_macros/src/symbols.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustc_macros/src/symbols.rs:152]: This PR is supposed to land only together with https://github.com/rust-lang/rust/pull/74554, which also regresses performance due to inability to do range checks, but improves it more due to inlining.

File: src/librustc_macros/src/symbols.rs [hunk 2]
```diff
@@ -44,22 +46,41 @@ impl Parse for Symbol {
...
+struct Keywords(IndexMap<Option<Ident>, Vec<Keyword>>);
 
-impl<T: Parse> Parse for List<T> {
+impl Parse for Keywords {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let mut classes = IndexMap::<Option<Ident>, Vec<Keyword>>::new();
+        let mut current_class = None;
+        while !input.is_empty() {
+            if input.peek(Token![fn]) {
+                input.parse::<TokenTree>()?;
```
Comment (petrochenkov) [inline src/librustc_macros/src/symbols.rs:58]: I wonder if there's any simpler way to skip a token with `syn` parser.

**Comment by petrochenkov** [comment]
Ok, `match` looks better than `contains`.

File: src/librustc_span/symbol.rs [hunk 3]
```diff
@@ -26,12 +26,14 @@ symbols! {
     Keywords {
         // Special reserved identifiers used internally for elided lifetimes,
         // unnamed method parameters, crate root module, error recovery etc.
+        fn is_special:
```
Comment (petrochenkov) [inline src/librustc_span/symbol.rs:29]: This confuses code highlighting in my editor (vscode + rust-analyzer), but `fn` is required for searchability.
Perhaps the list needs to go into braces, e.g. like this:
```rust
fn is_special {
    Invalid:            "",
    PathRoot:           "{{root}}",
    DollarCrate:        "$crate",
    Underscore:         "_",
}
```

**Comment by petrochenkov** [comment]
```
error[E0107]: wrong number of type arguments: expected 3, found 2
  --> src/librustc_macros/src/symbols.rs:50:17
   |
50 | struct Keywords(IndexMap<Option<Ident>, Vec<Keyword>>);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 type arguments
```
Excuse me?
Somehow `rustc_macros` gets built without `has_std` feature being enabled for the `indexmap` crate.

**Comment by Mark-Simulacrum** [comment]
Hm so indexmap is using autocfg to probe for std's presence, I think via this [bit of code](https://docs.rs/autocfg/1.0.0/src/autocfg/lib.rs.html#268-271). My guess is that's not using the right rustc or not passing it the right parameters, but I'm not sure without seeing the actual build log from autocfg we'd be able to say much.

**Comment by petrochenkov** [comment]
I couldn't reproduce the indexmap error locally, but fixed it by providing the hasher argument explicitly.

**Comment by shepmaster** [comment]
> My guess is that's not using the right rustc or not passing it the right parameters

See also https://github.com/rust-lang/cargo/issues/8646

**Comment by petrochenkov** [comment]
So, this PR should probably just be cherry-picked into https://github.com/rust-lang/rust/pull/74554 since it's not valuable by itself.

**Comment by petrochenkov** [comment]
Closing since https://github.com/rust-lang/rust/pull/74554 has been closed.

**Pull request closed without merging**